Pandas ia an essential library to learn if you are starting a career in machine learning. In this tuitorial I will go with basic use of pandas to advance. From installing to many important function of Pandas.

Let's get started.
https://www.kaggle.com/code/ahmedmohammedlethy/walmart-sales-forecasting/data?select=train.csv

# 1. Installation of pandas

From your terminal(Mac) or command line(PC) install it using:

### conda install pandas

or

### pip install pandas

If you are in jupyter notebook then you can install using:

### !pip install pandas

! just described that you where in terminal

In [1]:
#We will import our pandas library as:
import pandas as pd

# 2. Creating data from scratch
Pandas work with two data structure:
* Series
* DataFrame

## a. Series:
Series is a one-dimensional array object which you can say as a single column. The label asscoiated with column is called index.

In [2]:
# it shows index on left and values on right.
# index start from 0 to N-1 (where N is the length of the data)
obj = pd.Series([-2,9,7,1])
obj

0   -2
1    9
2    7
3    1
dtype: int64

In [3]:
# to get array representaion of values and index
obj.values

array([-2,  9,  7,  1])

In [4]:
obj.index

RangeIndex(start=0, stop=4, step=1)

In [5]:
# if you don't want default index, you can add your own label to index
obj1 = pd.Series([-2,9,7,1], index=['e','a','b','d'])
obj1

e   -2
a    9
b    7
d    1
dtype: int64

You can even create a series through dictionary:

In [6]:
data = {'Burlington': 3000, 'Walmart': 4500,'Costco': 10000,'Stop & Shop':2000}
obj2 = pd.Series(data)
obj2

Burlington      3000
Walmart         4500
Costco         10000
Stop & Shop     2000
dtype: int64

In [7]:
data = {'Burlington': 'abcd', 'Walmart': 4500,'Costco': 10000,'Stop & Shop':2000}
obj2 = pd.Series(data)
obj2

Burlington      abcd
Walmart         4500
Costco         10000
Stop & Shop     2000
dtype: object

## b. DataFrame:
DataFrame is a nothing but a collection of Series or a collection of columns. It is a two-dimensional structure which consist of rows and columns.

In [8]:
data = {'stores': ['Burlington','Walmart', 'Costco', 'Stop & Shop'] , 
        'states': ['IN', 'MA','CA','NJ'],
        'sales': [100000, 40000, 70000,50000]}
frame = pd.DataFrame(data)
frame

stores states   sales
0   Burlington     IN  100000
1      Walmart     MA   40000
2       Costco     CA   70000
3  Stop & Shop     NJ   50000

you can also give your own indexes if you want instead of default, you can simply do that by adding your index name:

In [9]:
data = {'stores': ['Burlington','Walmart', 'Costco', 'Stop & Shop'] , 
        'states': ['IN', 'MA','CA','NJ'],
        'sales': [100000, 40000, 70000,50000]}
frame = pd.DataFrame(data, index=['one','two','three','four'])
frame

stores states   sales
one     Burlington     IN  100000
two        Walmart     MA   40000
three       Costco     CA   70000
four   Stop & Shop     NJ   50000

# 3. Reading data files
We always don't create data manually with Series and DataFrame. We use already existing ddata. Data can be stored in any format: CSV, HTML, JSON, excel, SQL, etc. Below are some of the ways you can read the data:

Lets read our data. I have downloaded Walmart Sales Forecasting dataset from kaggle. I have downloaded two tables: stores.csv and train.csv
So, let's call that two dataset

In [10]:
df1 = pd.read_csv('stores.csv')
df1.head()

Store Type    Size
0      1    A  151315
1      2    A  202307
2      3    B   37392
3      4    A  205863
4      5    B   34875

In [11]:
df2 = pd.read_csv('train.csv')
df2.head()

Store  Dept        Date  Weekly_Sales  IsHoliday
0      1     1  2010-02-05      24924.50      False
1      1     1  2010-02-12      46039.49       True
2      1     1  2010-02-19      41595.55      False
3      1     1  2010-02-26      19403.54      False
4      1     1  2010-03-05      21827.90      False

In [12]:
df1.shape

(45, 3)

In [13]:
df2.shape

(421570, 5)

As you can see our stores dataset : df1 is of shape 45 rows and 3 columns and our train dataset: df2 is of shape 421570 rows and 5 columns. But you can see that both dataset has Store column in common. So, we can merge two dataset into one with a common column of Store.

* merge(): It is used to combine DataFarmes side by side based on common column.

In [14]:
final_df = df1.merge(df2, on="Store")

In [15]:
final_df.tail()

Store Type    Size  Dept        Date  Weekly_Sales  IsHoliday
421565     45    B  118221    98  2012-09-28        508.37      False
421566     45    B  118221    98  2012-10-05        628.10      False
421567     45    B  118221    98  2012-10-12       1061.02      False
421568     45    B  118221    98  2012-10-19        760.01      False
421569     45    B  118221    98  2012-10-26       1076.80      False

In [16]:
final_df.shape

(421570, 7)

Now, we have merge two dataset and we have 421570 rows and 7 columns. In above picture I have introduces tail function
#### tail(): It displays the last 5 rows of the dataset.

In [17]:
final_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 421570 entries, 0 to 421569
Data columns (total 7 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   Store         421570 non-null  int64  
 1   Type          421570 non-null  object 
 2   Size          421570 non-null  int64  
 3   Dept          421570 non-null  int64  
 4   Date          421570 non-null  object 
 5   Weekly_Sales  421570 non-null  float64
 6   IsHoliday     421570 non-null  bool   
dtypes: bool(1), float64(1), int64(3), object(2)
memory usage: 22.9+ MB


In [18]:
final_df.describe()

Store           Size           Dept   Weekly_Sales
count  421570.000000  421570.000000  421570.000000  421570.000000
mean       22.200546  136727.915739      44.260317   15981.258123
std        12.785297   60980.583328      30.492054   22711.183519
min         1.000000   34875.000000       1.000000   -4988.940000
25%        11.000000   93638.000000      18.000000    2079.650000
50%        22.000000  140167.000000      37.000000    7612.030000
75%        33.000000  202505.000000      74.000000   20205.852500
max        45.000000  219622.000000      99.000000  693099.360000

In [19]:
final_df.columns

Index(['Store', 'Type', 'Size', 'Dept', 'Date', 'Weekly_Sales', 'IsHoliday'], dtype='object')

# 4. Renaming the columns

In [20]:
final_df.rename(columns={'Store': 'store' ,'IsHoliday': 'is_holiday'}, inplace=True)
final_df.head()

store Type    Size  Dept        Date  Weekly_Sales  is_holiday
0      1    A  151315     1  2010-02-05      24924.50       False
1      1    A  151315     1  2010-02-12      46039.49        True
2      1    A  151315     1  2010-02-19      41595.55       False
3      1    A  151315     1  2010-02-26      19403.54       False
4      1    A  151315     1  2010-03-05      21827.90       False

In [21]:
# simplest method is: if you have to clean all column and make it lower case then:
final_df.columns = [x.lower() for x in final_df]
final_df.columns

Index(['store', 'type', 'size', 'dept', 'date', 'weekly_sales', 'is_holiday'], dtype='object')

In [22]:
final_df.head(2)

store type    size  dept        date  weekly_sales  is_holiday
0      1    A  151315     1  2010-02-05      24924.50       False
1      1    A  151315     1  2010-02-12      46039.49        True

In [23]:
#renaming using str.lower
final_df = final_df.rename(columns=str.lower)
final_df.head(2)

store type    size  dept        date  weekly_sales  is_holiday
0      1    A  151315     1  2010-02-05      24924.50       False
1      1    A  151315     1  2010-02-12      46039.49        True

# 5. To check null values and work with missing values

In [25]:
#no null values
final_df.isnull().sum()

store           0
type            0
size            0
dept            0
date            0
weekly_sales    0
is_holiday      0
dtype: int64

In [31]:
final_df['size'].unique()

array([151315, 202307,  37392, 205863,  34875, 202505,  70713, 155078,
       125833, 126512, 207499, 112238, 219622, 200898, 123737,  57197,
        93188, 120653, 203819, 203742, 140167, 119557, 114533, 128107,
       152513, 204184, 206302,  93638,  42988, 203750, 203007,  39690,
       158114, 103681,  39910, 184109, 155083, 196321,  41062, 118221])

In [33]:
final_df['store'].unique()

array([ 1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
       18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34,
       35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45])

In [34]:
#we replaced two first and last index with null to understand what to do when we have null values
import numpy as np
final_df.replace([24924.50, 1076.80], np.nan, inplace=True)

In [35]:
final_df

store type    size  dept        date  weekly_sales  is_holiday
0           1    A  151315     1  2010-02-05           NaN       False
1           1    A  151315     1  2010-02-12      46039.49        True
2           1    A  151315     1  2010-02-19      41595.55       False
3           1    A  151315     1  2010-02-26      19403.54       False
4           1    A  151315     1  2010-03-05      21827.90       False
...       ...  ...     ...   ...         ...           ...         ...
421565     45    B  118221    98  2012-09-28        508.37       False
421566     45    B  118221    98  2012-10-05        628.10       False
421567     45    B  118221    98  2012-10-12       1061.02       False
421568     45    B  118221    98  2012-10-19        760.01       False
421569     45    B  118221    98  2012-10-26           NaN       False

[421570 rows x 7 columns]

In [36]:
final_df.isnull().sum()

store           0
type            0
size            0
dept            0
date            0
weekly_sales    2
is_holiday      0
dtype: int64

## dropna(how="all"): it won't change anyting as our dataset has no rows with all NULL values

In [37]:

final_df.dropna(how="all",inplace=True)

In [38]:
final_df

store type    size  dept        date  weekly_sales  is_holiday
0           1    A  151315     1  2010-02-05           NaN       False
1           1    A  151315     1  2010-02-12      46039.49        True
2           1    A  151315     1  2010-02-19      41595.55       False
3           1    A  151315     1  2010-02-26      19403.54       False
4           1    A  151315     1  2010-03-05      21827.90       False
...       ...  ...     ...   ...         ...           ...         ...
421565     45    B  118221    98  2012-09-28        508.37       False
421566     45    B  118221    98  2012-10-05        628.10       False
421567     45    B  118221    98  2012-10-12       1061.02       False
421568     45    B  118221    98  2012-10-19        760.01       False
421569     45    B  118221    98  2012-10-26           NaN       False

[421570 rows x 7 columns]

## backward fill

In [39]:

final_df.fillna(method='bfill')

store type    size  dept        date  weekly_sales  is_holiday
0           1    A  151315     1  2010-02-05      46039.49       False
1           1    A  151315     1  2010-02-12      46039.49        True
2           1    A  151315     1  2010-02-19      41595.55       False
3           1    A  151315     1  2010-02-26      19403.54       False
4           1    A  151315     1  2010-03-05      21827.90       False
...       ...  ...     ...   ...         ...           ...         ...
421565     45    B  118221    98  2012-09-28        508.37       False
421566     45    B  118221    98  2012-10-05        628.10       False
421567     45    B  118221    98  2012-10-12       1061.02       False
421568     45    B  118221    98  2012-10-19        760.01       False
421569     45    B  118221    98  2012-10-26           NaN       False

[421570 rows x 7 columns]

## forward fill

In [40]:

final_df.fillna(method='ffill')

store type    size  dept        date  weekly_sales  is_holiday
0           1    A  151315     1  2010-02-05           NaN       False
1           1    A  151315     1  2010-02-12      46039.49        True
2           1    A  151315     1  2010-02-19      41595.55       False
3           1    A  151315     1  2010-02-26      19403.54       False
4           1    A  151315     1  2010-03-05      21827.90       False
...       ...  ...     ...   ...         ...           ...         ...
421565     45    B  118221    98  2012-09-28        508.37       False
421566     45    B  118221    98  2012-10-05        628.10       False
421567     45    B  118221    98  2012-10-12       1061.02       False
421568     45    B  118221    98  2012-10-19        760.01       False
421569     45    B  118221    98  2012-10-26        760.01       False

[421570 rows x 7 columns]

## fillna using mean method

In [41]:

mean_weekly_sales = final_df['weekly_sales'].mean().round(2)

In [42]:
mean_weekly_sales

15981.27

In [43]:
final_df.fillna(mean_weekly_sales, inplace=True)

In [44]:
final_df

store type    size  dept        date  weekly_sales  is_holiday
0           1    A  151315     1  2010-02-05      15981.27       False
1           1    A  151315     1  2010-02-12      46039.49        True
2           1    A  151315     1  2010-02-19      41595.55       False
3           1    A  151315     1  2010-02-26      19403.54       False
4           1    A  151315     1  2010-03-05      21827.90       False
...       ...  ...     ...   ...         ...           ...         ...
421565     45    B  118221    98  2012-09-28        508.37       False
421566     45    B  118221    98  2012-10-05        628.10       False
421567     45    B  118221    98  2012-10-12       1061.02       False
421568     45    B  118221    98  2012-10-19        760.01       False
421569     45    B  118221    98  2012-10-26      15981.27       False

[421570 rows x 7 columns]

# 6. To look for duplicates

In [45]:
# no duplicates in our dataframe
final_df.duplicated()

0         False
1         False
2         False
3         False
4         False
          ...  
421565    False
421566    False
421567    False
421568    False
421569    False
Length: 421570, dtype: bool

In [46]:
final_df.drop_duplicates()

store type    size  dept        date  weekly_sales  is_holiday
0           1    A  151315     1  2010-02-05      15981.27       False
1           1    A  151315     1  2010-02-12      46039.49        True
2           1    A  151315     1  2010-02-19      41595.55       False
3           1    A  151315     1  2010-02-26      19403.54       False
4           1    A  151315     1  2010-03-05      21827.90       False
...       ...  ...     ...   ...         ...           ...         ...
421565     45    B  118221    98  2012-09-28        508.37       False
421566     45    B  118221    98  2012-10-05        628.10       False
421567     45    B  118221    98  2012-10-12       1061.02       False
421568     45    B  118221    98  2012-10-19        760.01       False
421569     45    B  118221    98  2012-10-26      15981.27       False

[421570 rows x 7 columns]

# 7. Adding a column in a data frame and using of mapping method

In [47]:
# adding new row month and year
final_df['month'] = pd.DatetimeIndex(final_df['date']).month
final_df['year'] = pd.DatetimeIndex(final_df['date']).year
final_df.tail()

store type    size  dept        date  weekly_sales  is_holiday  month  \
421565     45    B  118221    98  2012-09-28        508.37       False      9   
421566     45    B  118221    98  2012-10-05        628.10       False     10   
421567     45    B  118221    98  2012-10-12       1061.02       False     10   
421568     45    B  118221    98  2012-10-19        760.01       False     10   
421569     45    B  118221    98  2012-10-26      15981.27       False     10   

        year  
421565  2012  
421566  2012  
421567  2012  
421568  2012  
421569  2012

In [48]:
final_df

store type    size  dept        date  weekly_sales  is_holiday  month  \
0           1    A  151315     1  2010-02-05      15981.27       False      2   
1           1    A  151315     1  2010-02-12      46039.49        True      2   
2           1    A  151315     1  2010-02-19      41595.55       False      2   
3           1    A  151315     1  2010-02-26      19403.54       False      2   
4           1    A  151315     1  2010-03-05      21827.90       False      3   
...       ...  ...     ...   ...         ...           ...         ...    ...   
421565     45    B  118221    98  2012-09-28        508.37       False      9   
421566     45    B  118221    98  2012-10-05        628.10       False     10   
421567     45    B  118221    98  2012-10-12       1061.02       False     10   
421568     45    B  118221    98  2012-10-19        760.01       False     10   
421569     45    B  118221    98  2012-10-26      15981.27       False     10   

        year  
0       2010  
1       2010  
2       2010  
3       2010  
4       2010  
...      ...  
421565  2012  
421566  2012  
421567  2012  
421568  2012  
421569  2012  

[421570 rows x 9 columns]

In [49]:
#for mapping method creating dictionary
month_nummber_to_name = {
      1: 'January',
      2: 'February',
      3: 'March',
      4: 'April',
      5: 'May',
      6: 'June',
      7: 'July',
      8: 'August',
      9: 'September',
      10: 'October',
      11: 'November',
      12: 'December'
}

In [50]:
#created new column month_name using mapping method
final_df['month_name'] = final_df['month'].map(month_nummber_to_name)

In [51]:
final_df.tail()

store type    size  dept        date  weekly_sales  is_holiday  month  \
421565     45    B  118221    98  2012-09-28        508.37       False      9   
421566     45    B  118221    98  2012-10-05        628.10       False     10   
421567     45    B  118221    98  2012-10-12       1061.02       False     10   
421568     45    B  118221    98  2012-10-19        760.01       False     10   
421569     45    B  118221    98  2012-10-26      15981.27       False     10   

        year month_name  
421565  2012  September  
421566  2012    October  
421567  2012    October  
421568  2012    October  
421569  2012    October

# 8. Slicing, selecting & extracting of dataframe

In [52]:
#column extracting
monthh = final_df[['month']]
monthh.head()

month
0      2
1      2
2      2
3      2
4      3

In [53]:
valuess = final_df[['month','date','is_holiday']]
valuess.head()

month        date  is_holiday
0      2  2010-02-05       False
1      2  2010-02-12        True
2      2  2010-02-19       False
3      2  2010-02-26       False
4      3  2010-03-05       False

In [54]:
type(monthh)

pandas.core.frame.DataFrame

In [55]:
final_df

store type    size  dept        date  weekly_sales  is_holiday  month  \
0           1    A  151315     1  2010-02-05      15981.27       False      2   
1           1    A  151315     1  2010-02-12      46039.49        True      2   
2           1    A  151315     1  2010-02-19      41595.55       False      2   
3           1    A  151315     1  2010-02-26      19403.54       False      2   
4           1    A  151315     1  2010-03-05      21827.90       False      3   
...       ...  ...     ...   ...         ...           ...         ...    ...   
421565     45    B  118221    98  2012-09-28        508.37       False      9   
421566     45    B  118221    98  2012-10-05        628.10       False     10   
421567     45    B  118221    98  2012-10-12       1061.02       False     10   
421568     45    B  118221    98  2012-10-19        760.01       False     10   
421569     45    B  118221    98  2012-10-26      15981.27       False     10   

        year month_name  
0       2010   February  
1       2010   February  
2       2010   February  
3       2010   February  
4       2010      March  
...      ...        ...  
421565  2012  September  
421566  2012    October  
421567  2012    October  
421568  2012    October  
421569  2012    October  

[421570 rows x 10 columns]

## row extraction using loc and iloc

In [56]:
final_df.loc[1,['month_name','weekly_sales','date']]

month_name        February
weekly_sales      46039.49
date            2010-02-12
Name: 1, dtype: object

In [57]:
final_df.iloc[1,[9,5,4]]

month_name        February
weekly_sales      46039.49
date            2010-02-12
Name: 1, dtype: object

## selecting a range of rows using loc and iloc using slicing:

In [58]:

final_df.loc[2:5]

store type    size  dept        date  weekly_sales  is_holiday  month  \
2      1    A  151315     1  2010-02-19      41595.55       False      2   
3      1    A  151315     1  2010-02-26      19403.54       False      2   
4      1    A  151315     1  2010-03-05      21827.90       False      3   
5      1    A  151315     1  2010-03-12      21043.39       False      3   

   year month_name  
2  2010   February  
3  2010   February  
4  2010      March  
5  2010      March

In [59]:
final_df.iloc[2:5]

store type    size  dept        date  weekly_sales  is_holiday  month  \
2      1    A  151315     1  2010-02-19      41595.55       False      2   
3      1    A  151315     1  2010-02-26      19403.54       False      2   
4      1    A  151315     1  2010-03-05      21827.90       False      3   

   year month_name  
2  2010   February  
3  2010   February  
4  2010      March

## Selecting data based on some condition:

In [60]:
final_df.loc[(final_df['month_name']=='March') & (final_df['weekly_sales']>=80000)]

store type    size  dept        date  weekly_sales  is_holiday  month  \
5122        1    A  151315    38  2010-03-05     108282.86       False      3   
5123        1    A  151315    38  2010-03-12      98175.35       False      3   
5125        1    A  151315    38  2010-03-26      95855.41       False      3   
5174        1    A  151315    38  2011-03-04      89649.84       False      3   
5226        1    A  151315    38  2012-03-02      93103.43       False      3   
...       ...  ...     ...   ...         ...           ...         ...    ...   
403867     43    C   41062    92  2012-03-02      90232.68       False      3   
403868     43    C   41062    92  2012-03-09      90195.91       False      3   
403869     43    C   41062    92  2012-03-16      91226.88       False      3   
403870     43    C   41062    92  2012-03-23      80545.44       False      3   
404300     43    C   41062    95  2012-03-30      94764.38       False      3   

        year month_name  
5122    2010      March  
5123    2010      March  
5125    2010      March  
5174    2011      March  
5226    2012      March  
...      ...        ...  
403867  2012      March  
403868  2012      March  
403869  2012      March  
403870  2012      March  
404300  2012      March  

[916 rows x 10 columns]

In [61]:
final_df.iloc[list((final_df['month_name']=='March') & (final_df['weekly_sales']>=80000))]

store type    size  dept        date  weekly_sales  is_holiday  month  \
5122        1    A  151315    38  2010-03-05     108282.86       False      3   
5123        1    A  151315    38  2010-03-12      98175.35       False      3   
5125        1    A  151315    38  2010-03-26      95855.41       False      3   
5174        1    A  151315    38  2011-03-04      89649.84       False      3   
5226        1    A  151315    38  2012-03-02      93103.43       False      3   
...       ...  ...     ...   ...         ...           ...         ...    ...   
403867     43    C   41062    92  2012-03-02      90232.68       False      3   
403868     43    C   41062    92  2012-03-09      90195.91       False      3   
403869     43    C   41062    92  2012-03-16      91226.88       False      3   
403870     43    C   41062    92  2012-03-23      80545.44       False      3   
404300     43    C   41062    95  2012-03-30      94764.38       False      3   

        year month_name  
5122    2010      March  
5123    2010      March  
5125    2010      March  
5174    2011      March  
5226    2012      March  
...      ...        ...  
403867  2012      March  
403868  2012      March  
403869  2012      March  
403870  2012      March  
404300  2012      March  

[916 rows x 10 columns]

# 9. isin() method:

In [62]:
filter1 = final_df['month_name'].isin(['February','May'])
filter2 = final_df['store'].isin([10,21,43])
final_df[filter1 & filter2]

store type    size  dept        date  weekly_sales  is_holiday  month  \
87524      10    B  126512     1  2010-02-05      40212.84       False      2   
87525      10    B  126512     1  2010-02-12      67699.32        True      2   
87526      10    B  126512     1  2010-02-19      49748.33       False      2   
87527      10    B  126512     1  2010-02-26      33601.22       False      2   
87537      10    B  126512     1  2010-05-07      29819.93       False      5   
...       ...  ...     ...   ...         ...           ...         ...    ...   
404734     43    C   41062    98  2012-05-04       8465.30       False      5   
404735     43    C   41062    98  2012-05-11      10618.13       False      5   
404736     43    C   41062    98  2012-05-18      10646.37       False      5   
404737     43    C   41062    98  2012-05-25       8179.45       False      5   
404760     43    C   41062    99  2011-05-06         25.00       False      5   

        year month_name  
87524   2010   February  
87525   2010   February  
87526   2010   February  
87527   2010   February  
87537   2010        May  
...      ...        ...  
404734  2012        May  
404735  2012        May  
404736  2012        May  
404737  2012        May  
404760  2011        May  

[4472 rows x 10 columns]

In [63]:
final_df

store type    size  dept        date  weekly_sales  is_holiday  month  \
0           1    A  151315     1  2010-02-05      15981.27       False      2   
1           1    A  151315     1  2010-02-12      46039.49        True      2   
2           1    A  151315     1  2010-02-19      41595.55       False      2   
3           1    A  151315     1  2010-02-26      19403.54       False      2   
4           1    A  151315     1  2010-03-05      21827.90       False      3   
...       ...  ...     ...   ...         ...           ...         ...    ...   
421565     45    B  118221    98  2012-09-28        508.37       False      9   
421566     45    B  118221    98  2012-10-05        628.10       False     10   
421567     45    B  118221    98  2012-10-12       1061.02       False     10   
421568     45    B  118221    98  2012-10-19        760.01       False     10   
421569     45    B  118221    98  2012-10-26      15981.27       False     10   

        year month_name  
0       2010   February  
1       2010   February  
2       2010   February  
3       2010   February  
4       2010      March  
...      ...        ...  
421565  2012  September  
421566  2012    October  
421567  2012    October  
421568  2012    October  
421569  2012    October  

[421570 rows x 10 columns]

In [64]:
final_df[['is_holiday']]

is_holiday
0            False
1             True
2            False
3            False
4            False
...            ...
421565       False
421566       False
421567       False
421568       False
421569       False

[421570 rows x 1 columns]

In [65]:
final_df[final_df['month'] == 2]

store type    size  dept        date  weekly_sales  is_holiday  month  \
0           1    A  151315     1  2010-02-05      15981.27       False      2   
1           1    A  151315     1  2010-02-12      46039.49        True      2   
2           1    A  151315     1  2010-02-19      41595.55       False      2   
3           1    A  151315     1  2010-02-26      19403.54       False      2   
52          1    A  151315     1  2011-02-04      21665.76       False      2   
...       ...  ...     ...   ...         ...           ...         ...    ...   
421482     45    B  118221    98  2011-02-25        675.26       False      2   
421531     45    B  118221    98  2012-02-03        772.51       False      2   
421532     45    B  118221    98  2012-02-10        770.86        True      2   
421533     45    B  118221    98  2012-02-17        889.98       False      2   
421534     45    B  118221    98  2012-02-24        544.34       False      2   

        year month_name  
0       2010   February  
1       2010   February  
2       2010   February  
3       2010   February  
52      2011   February  
...      ...        ...  
421482  2011   February  
421531  2012   February  
421532  2012   February  
421533  2012   February  
421534  2012   February  

[35526 rows x 10 columns]

# 9. sort_values() method:

In [66]:
final_df.sort_values(by='weekly_sales')

store type    size  dept        date  weekly_sales  is_holiday  month  \
267730     28    A  206302     6  2010-10-08      -4988.94       False     10   
336495     35    B  103681    47  2011-09-02      -3924.00       False      9   
417801     45    B  118221    47  2010-02-26      -1750.00       False      2   
153916     16    B   57197    47  2012-03-09      -1699.00       False      3   
271300     28    A  206302    32  2011-03-25      -1321.48       False      3   
...       ...  ...     ...   ...         ...           ...         ...    ...   
135665     14    A  200898    72  2010-11-26     474330.10        True     11   
337961     35    B  103681    72  2010-11-26     627962.93        True     11   
95425      10    B  126512    72  2011-11-25     630999.19        True     11   
338013     35    B  103681    72  2011-11-25     649770.18        True     11   
95373      10    B  126512    72  2010-11-26     693099.36        True     11   

        year month_name  
267730  2010    October  
336495  2011  September  
417801  2010   February  
153916  2012      March  
271300  2011      March  
...      ...        ...  
135665  2010   November  
337961  2010   November  
95425   2011   November  
338013  2011   November  
95373   2010   November  

[421570 rows x 10 columns]

In [71]:
final_df.loc[final_df['month_name'] == 'January'].sort_values(by='month')

store type    size  dept        date  weekly_sales  is_holiday  month  \
48          1    A  151315     1  2011-01-07      15984.24       False      1   
280684     29    B   93638    27  2012-01-27        851.09       False      1   
280683     29    B   93638    27  2012-01-20       1443.75       False      1   
280682     29    B   93638    27  2012-01-13       1448.51       False      1   
280681     29    B   93638    27  2012-01-06       1555.93       False      1   
...       ...  ...     ...   ...         ...           ...         ...    ...   
140172     15    B  123737    16  2011-01-21       2689.82       False      1   
140171     15    B  123737    16  2011-01-14       1989.29       False      1   
140170     15    B  123737    16  2011-01-07       1906.77       False      1   
140081     15    B  123737    14  2012-01-20       9925.36       False      1   
421530     45    B  118221    98  2012-01-27        671.40       False      1   

        year month_name  
48      2011    January  
280684  2012    January  
280683  2012    January  
280682  2012    January  
280681  2012    January  
...      ...        ...  
140172  2011    January  
140171  2011    January  
140170  2011    January  
140081  2012    January  
421530  2012    January  

[23545 rows x 10 columns]

In [75]:
final_df.loc[final_df['month_name'] == 'January'].sort_values(by='weekly_sales')

store type    size  dept        date  weekly_sales  is_holiday  month  \
113824     12    B  112238    47  2011-01-28       -400.00       False      1   
94292      10    B  126512    54  2012-01-20       -370.12       False      1   
336536     35    B  103681    49  2012-01-20       -326.61       False      1   
326509     34    A  158114    47  2011-01-07       -224.00       False      1   
54845       6    A  202505    47  2012-01-20       -119.00       False      1   
...       ...  ...     ...   ...         ...           ...         ...    ...   
38848       4    A  205863    92  2012-01-13     186643.74       False      1   
196680     20    A  203742    92  2011-01-14     190728.60       False      1   
196732     20    A  203742    92  2012-01-13     197649.89       False      1   
137254     14    A  200898    92  2011-01-07     198789.07       False      1   
137306     14    A  200898    92  2012-01-06     206871.52       False      1   

        year month_name  
113824  2011    January  
94292   2012    January  
336536  2012    January  
326509  2011    January  
54845   2012    January  
...      ...        ...  
38848   2012    January  
196680  2011    January  
196732  2012    January  
137254  2011    January  
137306  2012    January  

[23545 rows x 10 columns]

In [77]:
#sorthing in descending order
final_df.loc[final_df['month_name'] == 'January'].sort_values(by=['weekly_sales'], ascending=False)

store type    size  dept        date  weekly_sales  is_holiday  month  \
137306     14    A  200898    92  2012-01-06     206871.52       False      1   
137254     14    A  200898    92  2011-01-07     198789.07       False      1   
196732     20    A  203742    92  2012-01-13     197649.89       False      1   
196680     20    A  203742    92  2011-01-14     190728.60       False      1   
38848       4    A  205863    92  2012-01-13     186643.74       False      1   
...       ...  ...     ...   ...         ...           ...         ...    ...   
54845       6    A  202505    47  2012-01-20       -119.00       False      1   
326509     34    A  158114    47  2011-01-07       -224.00       False      1   
336536     35    B  103681    49  2012-01-20       -326.61       False      1   
94292      10    B  126512    54  2012-01-20       -370.12       False      1   
113824     12    B  112238    47  2011-01-28       -400.00       False      1   

        year month_name  
137306  2012    January  
137254  2011    January  
196732  2012    January  
196680  2011    January  
38848   2012    January  
...      ...        ...  
54845   2012    January  
326509  2011    January  
336536  2012    January  
94292   2012    January  
113824  2011    January  

[23545 rows x 10 columns]

# 10. value_counts() method:

In [101]:
final_df.value_counts(['is_holiday'])

is_holiday
False         391909
True           29661
dtype: int64

In [110]:
final_df.value_counts(['year']).sort_index()

year
2010    140679
2011    153453
2012    127438
dtype: int64

In [86]:
final_df.value_counts(['dept']).sort_index()

dept
1       6435
2       6435
3       6435
4       6435
5       6347
        ... 
95      6435
96      4854
97      6278
98      5836
99       862
Length: 81, dtype: int64